### ISSUE TITLE
Hot Reload test scenario: Hot Reload notifications and reloadable types

### ISSUE THREAD

**Issue by tmat** [dotnet/roslyn#55561]
1. Create .NET 6 Console Application with the following code:

```C#
using System.Reflection.Metadata;
using System.Runtime.CompilerServices;

[assembly: MetadataUpdateHandler(typeof(UpdateHandler))]

internal static class UpdateHandler
{

    public static void UpdateApplication(Type[] types)
    {
        Console.WriteLine("Types:");
        Console.WriteLine(types == null ? "<null>" : types.Length == 0 ? "<empty>" : string.Join(",", types.Select(t => t.Name)));
    }
}

[CreateNewOnMetadataUpdate]
class C
{
    public void F()
    {
        Console.Write("xyz");
    }
}

class Program
{
    static void Render()
    {
        Console.Write('.');
    }

    static void Main()
    {
        while (true)
        {
            Thread.Sleep(100);
            Render();
        }
    }
}
```

2. F5 or Ctrl+F5
3. Edit body of method C.F
4. Apply Changes 
![image](https://user-images.githubusercontent.com/41759/129119464-fef5e630-9a40-49dc-b536-2af11ece76de.png)
5. Program should output
```
Types:
C#1
```
6. Make another change and apply 
7. Program should output
```
Types:
C#2
```
8. Change `Console.Write('.');` to `Console.Write(':');` and Apply Changes
9. Program should start printing sequence of `:` instead of `.`

**Comment by tmat** [comment]
Currently doesn't work:
1) UpdateHandler is not triggered in F5 (fix merged on 8/10)
2) Types are empty

**Comment by tmat** [comment]
Should work since 17.0.
